sycl/include/CL/sycl/handler.hpp

```diff
@@ -898,6 +906,103 @@ class __SYCL_EXPORT handler {
     KernelFunc(detail::Builder::getElement(detail::declptr<ElementType>()));
   }
 
+  // NOTE: the name of this function - "kernel_parallel_for_work_group" - is
+  // used by the Front End to determine kernel invocation kind.
+  template <typename KernelName, typename ElementType, typename KernelType>
+  __attribute__((sycl_kernel)) void
+#ifdef __SYCL_NONCONST_FUNCTOR__
+  kernel_parallel_for_work_group(KernelType KernelFunc, kernel_handler KH) {
+#else
+  kernel_parallel_for_work_group(const KernelType &KernelFunc,
+                                 kernel_handler KH) {
+#endif
+    KernelFunc(detail::Builder::getElement(detail::declptr<ElementType>()), KH);
+  }
+
+  // Wrappers for kernel_*** functions above with and without support of
+  // additional kernel_handler argument.
+
+  // NOTE: to support kernel_handler argument in kernel lambdas, only
+  // kernel_***_wrapper functions must be called in this code
```

Comment:
> This condition - only kernel_***_wrapper functions must be called in this code - does not hold

Ca you please elaborate on this a bit? What I meant by this is that we shouldn't call, e.g., `kernel_parallel_for` in our headers anywhere because we need some logic above it to decide what version of `kernel_parallel_for` to run - with or w/o `kernel_handler` support. And we need some place to initialize `kernel_handler` obj on our side. In these cases `kernel_parallel_for_wrapper` acts as this place.

> Can it be a problem if the constructor becomes non-trivial?

It can if something will be changed in spec, in this case initialization of `kernel_handler` object should be re-written here. Right now we have separate init function `__init_specialization_constants_buffer` which is called in FE.

Reply:
OK, I see. This is not quite clear from the comments - I'd suggest to reword for better clarity.
My guess was that this was related more to parallel_for_work_group, where variables local to a kernel become WG-shared (address space 3 global in LLVM IR), and there is code filtering for the `WG leader` work item.